File: 4_Actors/actors/lib/hello_actors.ex
import :timer, only: [ sleep: 1 ]

defmodule Talker do

  def loop do
    receive do  # receive block waits for a message
      # Pattern matching
      {:greet, name} -> IO.puts("Hello #{name}")
      {:praise, name} -> IO.puts("#{name} you are amazing")
      {:celebrate, name, age} -> IO.puts("Here's to another #{age} years, #{name}")
    end
    loop()
  end
end



pid = spawn(&Talker.loop/0)
send(pid, {:greet, "Huey"})
send(pid, {:praise, "Dewey"})
send(pid, {:celebrate, "Louie", 12})
sleep(1000)

File: 4_Actors_Scala/actors/src/main/scala/hello_actors/HelloActors.scala
package org.woodtlilukas
package hello_actors

import akka.actor.{Actor, ActorRef, ActorSystem, PoisonPill, Props, Terminated}

case class Greet(name: String)
case class Praise(name: String)
case class Celebrate(name: String, age: Int)


class Talker extends Actor {

  def receive: Receive = {
    case Greet(name) => println(s"Hello $name")
    case Praise(name) => println(s"$name, you're amazing")
    case Celebrate(name, age) => println(s"Here's to another $age, $name")
  }
}


class Master extends Actor {

  val talker: ActorRef = context.actorOf(Props[Talker], "talker")

  override def preStart: Unit = {
    // establishing a death watch
    context.watch(talker)

    talker ! Greet("Huey")
    talker ! Praise("Dewey")
    talker ! Celebrate("Louie", 16)

    talker ! PoisonPill
  }

  def receive: Receive = {
    // `talker`: backticks to match value of existing talker variable
    case Terminated(`talker`) => context.system.terminate()
  }
}

object HelloActors extends App {

  val system = ActorSystem("HelloActors")

  // `Props` acts as a factory
  system.actorOf(Props[Master], "master")
}
